Upstream  0.13.8 + rest of R19

motobob opened the pull request:
upstream changes + All R19 compatibility spec fixes

mbystryantsev committed:
Get rid of compile-time warning about hash functions

mbystryantsev committed:
Provide last_modified value in get_object function results

mbystryantsev committed:
Provide next_marker attribute in list_objects function results

Commit:
Merge pull request #276 from horror-x/s3_improvements

Minor helpfull S3 improvements

okeuday committed:
-spec syntax fix for Erlang/OTP 19.0

okeuday committed:
Add eini application dependency.

Commit:
Merge pull request #280 from okeuday/add_eini_dependency

Add eini application dependency.

waisbrot committed:
Use Hex.pm for the eini dependency

Commit:
Merge pull request #281 from waisbrot/eini-hex

Use Hex.pm for the eini dependency

okeuday committed:
-include_lib for erlcloud is changed to -include to work with rebar2

Commit:
Add convenience function module erlcloud_ec2_meta to fetch items from AWS metadata, User Data and Dynamic Data
Add new features to erlcloud_as to describe lifecycle hooks and to complete lifecycle actions

Commit:
Merge pull request #279 from okeuday/fixes_for_19_0

Fixes for Erlang/OTP 19.0

timofey-barmin committed:
Make associate_public_ip optional in ec2 instance creation

If more than one interface is specified server returns
"The associatePublicIPAddress parameter cannot be specified when
launching with multiple network interfaces"

Commit:
Add support for elb attributes

Commit:
Return the specs back to the format used. That was fixed by #279

timofey-barmin committed:
security_group and private_ip should be [] by default in ec2_net_if

timofey-barmin committed:
Spec change: associate_public_ip could be undefined

Commit:
Merge pull request #286 from posilva/feature/describe_elb_attributes

Add support for elb attributes

Commit:
Merge pull request #285 from timofey-barmin/optional-public-ip-opt-in-net-list2

Make associate_public_ip optional in ec2 instance creation

okeuday committed:
Add compatibility source code for Erlang/OTP 19.0

* Switch to rand:uniform/1 instead of random:uniform/1 when using 19.x or higher

Commit:
Merge pull request #287 from okeuday/compat_for_19_0

Add compatibility source code for Erlang/OTP 19.0

Commit:
Implemented erlcloud_mon:get_metric_statistics for AWS CloudWatch

Commit:
Merge pull request #291 from samble/master

Implemented erlcloud_mon:get_metric_statistics for AWS CloudWatch

fogfish committed:
add security token to config

allow to define temporary security credentials to config object.

Commit:
Merge pull request #293 from fogfish/master

add security token to config

Commit:
add 19 to travis

Commit:
Merge pull request #294 from erlcloud/add_19

add 19 to travis

fogfish committed:
improve the logic of temporary credentials

Use profile feature to read security token instead of new api.

fogfish committed:
Merge remote-tracking branch 'upstream/master'

fogfish committed:
use undefined at token pattern match

Commit:
Merge pull request #295 from fogfish/master

add security token to profile

smilerlee committed:
Fix a bug of erlcloud_http:url_encode/1 according to RFC 3629

smilerlee committed:
Merge pull request #297 from erlcloud/url_encode

Fix a bug of erlcloud_http:url_encode/1 according to RFC 3629

juantascon committed:
add support for lhttpc pools

Commit:
Fix a bug in erlcloud_ec2:register_image/2

nalundgaard committed:
Merge pull request #299 from mtaylor91/master

Fix a bug in erlcloud_ec2:register_image/2

nalundgaard committed:
Merge pull request #300 from juantascon/lhttpc_pools

add support for lhttpc pools

ephe-meral committed:
ec2 describe_instances - correct spec of return val

Commit:
Merge pull request #301 from ephe-meral/patch-1

EC2 `describe_instances`: Corrects spec of return value

Commit:
Merge pull request #282 from ncrothe/master

Add function to fetch instance id

Commit:
merge up + rest of R19 fixes

motobob commented:
this one is to substitute #115 too

motobob commented on src/erlcloud.app.src:
```diff
@@ -12,6 +12,7 @@
                   ssl,
                   xmerl,
                   inets,
+                  eini,
```
manual merge error. to be removed

Commit:
fix app.src double eini

motobob commented:
test it

motobob commented:
@nalundgaard @noproc please take a look and merge

nalundgaard commented:
Holy big change, batman!

Looks like this is mostly the spec fixes for 19 and some very minor other new additions. I tested this with our internal functional testing for SES and DDB, and it looks good, so I am +1 there. I don't deal with any SW that touches the assumeRole stuff, but it looks like one area where there are some changes. Have you guys verified compatibility there?

Other than that, I am 👍 on this.

waisbrot committed:
allow undefined expiration

waisbrot committed:
Avoid empty headers in aws_request, allow empty response

waisbrot committed:
Erlang 19 dialyzer fixes for WAF and SNS

Commit:
Merge pull request #306 from waisbrot/dialyzer-cleanup-waf-sns

Erlang 19 dialyzer fixes for WAF and SNS

Commit:
Merge pull request #305 from waisbrot/dialyzer-cleanup-s3

Erlang 19 dialyzer fixes for S3

waisbrot committed:
Update eini package

Update to use the erlcloud/eini repo and the 1.2.2 tag

Commit:
Merge pull request #308 from waisbrot/update-eini

Update eini package

Commit:
cleanup couple gaps of previous merges

nalundgaard committed:
Merge pull request #309 from erlcloud/minor_cleanups

cleanup couple gaps of previous merges

Commit:
clean include_libs further

nalundgaard committed:
Merge pull request #310 from erlcloud/minor_cleanups

clean include_libs further

Commit:
mixup leftover specifica to AL fork

Commit:
bring up 0.13.8

except manual undo for erlcloud_aws.hrl aws_request{service} typing as
it is used more wider here.

motobob commented:
tested in real stack - ok.
tested the profile loading too - it works too. 

shipping it
